File: algorithms/medium/minimum_height_trees.ex
defmodule Solution do
  @spec find_min_height_trees(n :: integer, edges :: [[integer]]) :: [integer]
  def find_min_height_trees(1, _), do: [0]

  def find_min_height_trees(n, edges) do
    {deg, graph} = generate_graph(edges, %{}, %{})
    leaves = deg |> Enum.filter(fn {_, v} -> v == 1 end) |> Enum.map(fn {k, v} -> k end)
    last_leaves(n, leaves, deg, graph)
  end

  def last_leaves(n, leaves, _, _) when n <= 2, do: leaves

  def last_leaves(n, leaves, deg, graph) do
    {new_deg, new_leaves} = eliminate_leaves(leaves, graph, deg, [])
    last_leaves(n - Enum.count(leaves), new_leaves, new_deg, graph)
  end

  def eliminate_leaves([], graph, deg, leaves), do: {deg, leaves}

  def eliminate_leaves([leaf | rest], graph, deg, leaves) do
    {new_deg, new_leaves} = update_degrees(graph[leaf], deg, leaves)
    eliminate_leaves(rest, graph, new_deg, new_leaves)
  end

  def update_degrees([], deg, leaves), do: {deg, leaves}

  def update_degrees([n | rest], deg, leaves) do
    deg = Map.update(deg, n, 0, fn d -> d - 1 end)

    if deg[n] == 1 do
      update_degrees(rest, deg, [n | leaves])
    else
      update_degrees(rest, deg, leaves)
    end
  end

  def generate_graph([], degrees, graph) do
    {degrees, graph}
  end

  def generate_graph([[n1, n2] | rest], degrees, graph) do
    degrees =
      degrees
      |> Map.update(n1, 1, fn n -> n + 1 end)
      |> Map.update(n2, 1, fn n -> n + 1 end)

    graph =
      graph
      |> Map.update(n1, [n2], fn lst -> [n2 | lst] end)
      |> Map.update(n2, [n1], fn lst -> [n1 | lst] end)

    generate_graph(rest, degrees, graph)
  end
end

File: algorithms/medium/minimum_height_trees.rs
impl Solution {
    pub fn find_min_height_trees(mut n: i32, edges: Vec<Vec<i32>>) -> Vec<i32> {
        if n == 1 {
            return vec![0];
        }

        let mut adjacency_list = vec![vec![]; n as usize];
        let mut degrees = vec![0; n as usize];
        for edge in &edges {
            let n1 = edge[0] as usize;
            let n2 = edge[1] as usize;
            adjacency_list[n1].push(n2);
            adjacency_list[n2].push(n1);
            degrees[n1] += 1;
            degrees[n2] += 1;
        }

        let mut leaves: Vec<_> = degrees
            .iter()
            .enumerate()
            .filter(|&(_, d)| *d == 1)
            .map(|(i, _)| i)
            .collect();

        while n > 2 {
            n -= leaves.len() as i32;
            let mut new_leaves = Vec::new();
            for &leaf in &leaves {
                for &neighbor in &adjacency_list[leaf] {
                    degrees[neighbor] -= 1;
                    if degrees[neighbor] == 1 {
                        new_leaves.push(neighbor);
                    }
                }
            }
            leaves = new_leaves;
        }
        leaves.into_iter().map(|n| n as i32).collect()
    }
}
